Modix.Web/Pages/Promotions.razor

```diff
@@ -0,0 +1,348 @@
+                            }
+                        }
+                    </ChildContent>
+                </MudExpansionPanel>
+            }
+        </MudExpansionPanels>
+    </MudContainer>
+
+</CascadingAuthenticationState>
+
+<style>
+    .targetRole{
+        font-weight: 400 !important;
+        padding: 4px 8px;
+        border: 2px solid currentColor;
+        border-top-color: currentcolor;
+        border-right-color: currentcolor;
+        border-bottom-color: currentcolor;
+        border-left-color: currentcolor;
+        border-color: currentColor !important;
+        border-radius: 3px;
+        margin-left: .5em
+    }
+</style>
+
+@code {
+    [Inject]
+    public SessionState SessionState { get; set; } = null!;
+
+    [Inject]
+    public CookieService CookieService { get; set; } = null!;
+
+    [Inject]
+    public DiscordHelper DiscordHelper { get; set; } = null!;
+
+    [Inject]
+    public IPromotionsService PromotionsService { get; set; } = null!;
+
+    [Inject]
+    public IDialogService DialogService { get; set; } = null!;
+
+    [Inject]
+    public ISnackbar Snackbar { get; set; } = null!;
+
+    private ulong CurrentUserId { get; set; }
+
+    private IReadOnlyCollection<PromotionCampaignSummary> Campaigns = Array.Empty<PromotionCampaignSummary>();
+    private Dictionary<ulong, string> RoleColors = new Dictionary<ulong, string>();
+    private Dictionary<long, Dictionary<long, CampaignCommentData>> campaignCommentData = new Dictionary<long, Dictionary<long, CampaignCommentData>>();
+
+    private bool _showInactive;
+
+    protected override void OnInitialized()
+    {
+        _showInactive = SessionState.ShowInactivePromotions;
+    }
+
+    protected override async Task OnAfterRenderAsync(bool firstRender)
+    {
+        if (!firstRender)
+            return;
+
+        var currentUser = DiscordHelper.GetCurrentUser();
+        RoleColors = currentUser!.Guild.Roles.ToDictionary(x => x.Id, x => x.Color.ToString());
+
+        Campaigns = await PromotionsService.SearchCampaignsAsync(new PromotionCampaignSearchCriteria
+        {
+            GuildId = currentUser.Guild.Id
+        });
+
+        CurrentUserId = currentUser.Id;
+
+        StateHasChanged();
+    }
+
+    private async Task ShowInactiveChanged(bool showInactive)
+    {
+        _showInactive = showInactive;
+        await CookieService.SetShowInactivePromotionsAsync(showInactive);
+    }
+
+    private async Task CampaignExpanded(bool wasExpanded, long campaignId, ulong userId)
+    {
+        if (!wasExpanded)
+            return;
+
+        if (CurrentUserId == userId)
+            return;
+
+        if (campaignCommentData.ContainsKey(campaignId))
+            return;
+
+        var result = await PromotionsService.GetCampaignDetailsAsync(campaignId);
+
+        if (result is null) // TODO: What to do here?
```

Comment:
Not sure there's much we can do if the service can't find the campaign we're looking for. Early return like you have here is probably fine, otherwise maybe a slightly better option is to display an error message if possible.

Reply:
Added a snackbar message, not sure it'll ever happen but atleast it's there now :)